[iPad] Unable to mute the video on hulu.com

eric-carlson opened the pull request:
#### 43bf134a6b6ad911b061d17579fe3b02314f4cc5
<pre>
[iPad] Unable to mute the video on hulu.com
<a href="https://bugs.webkit.org/show_bug.cgi?id=281697">https://bugs.webkit.org/show_bug.cgi?id=281697</a>
<a href="https://rdar.apple.com/126096361">rdar://126096361</a>

Reviewed by Jer Noble.

Mute implicitly when volume is set to zero on hulu.com on iOS, where we don&apos;t have
per-element audio volume control, so their custom controls that use volume to mute and
unmute will work.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::HTMLMediaElement): Drive-by change, remove unused member
variable `m_initiallyMuted`.
(WebCore::HTMLMediaElement::setReadyState): Ditto.
(WebCore::HTMLMediaElement::volume const): Return 0 when implicitly muted.
(WebCore::HTMLMediaElement::setVolume): Mute implicitly when volume set to 0, clear
implicit mute when volume set to anything else.
(WebCore::HTMLMediaElement::muted const): Return true when implicitly muted.
(WebCore::HTMLMediaElement::setMuted): Pass flag to setMutedInternal so it runs the update
logic even when `m_muted` doesn&apos;t change bug implicit muted does.
(WebCore::HTMLMediaElement::setMutedInternal):
* Source/WebCore/html/HTMLMediaElement.h:

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::implicitMuteWhenVolumeSetToZero const): New Quirk.
* Source/WebCore/page/Quirks.h:

Canonical link: <a href="https://commits.webkit.org/285768@main">https://commits.webkit.org/285768@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3853b047bc9550be2de810935f0b78903111da67

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73756 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53185 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26567 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78078 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24994 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75873 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62318 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/970 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/58023 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16388 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76823 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48159 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63463 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38422 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44943 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20951 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23327 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66518 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21298 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79645 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1073 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/530 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66358 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1215 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63473 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/65637 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16216 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9502 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7686 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1037 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3787 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1066 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1053 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1072 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -4574,6 +4572,9 @@ void HTMLMediaElement::setControls(bool controls)
 
 double HTMLMediaElement::volume() const
 {
+    if (m_implicitlyMuted.value_or(false))
```
We do this pattern a lot: `m_implicitlyMuted.value_or(false)`. Could we just add a `bool implicitlyMuted() const { return m_implicitlyMuted.value_or(false); }` to HTMLMediaElement?

eric-carlson committed:
[iPad] Unable to mute the video on hulu.com
https://bugs.webkit.org/show_bug.cgi?id=281697
rdar://126096361

Reviewed by Jer Noble.

Mute implicitly when volume is set to zero on hulu.com on iOS, where we don't have
per-element audio volume control, so their custom controls that use volume to mute and
unmute will work.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::HTMLMediaElement): Drive-by change, remove unused member
variable `m_initiallyMuted`.
(WebCore::HTMLMediaElement::setReadyState): Ditto.
(WebCore::HTMLMediaElement::volume const): Return 0 when implicitly muted.
(WebCore::HTMLMediaElement::setVolume): Mute implicitly when volume set to 0, clear
implicit mute when volume set to anything else.
(WebCore::HTMLMediaElement::muted const): Return true when implicitly muted.
(WebCore::HTMLMediaElement::setMuted): Pass flag to setMutedInternal so it runs the update
logic even when `m_muted` doesn't change bug implicit muted does.
(WebCore::HTMLMediaElement::setMutedInternal):
* Source/WebCore/html/HTMLMediaElement.h:

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::implicitMuteWhenVolumeSetToZero const): New Quirk.
* Source/WebCore/page/Quirks.h:

Canonical link: https://commits.webkit.org/285768@main
